spec.md

```diff
@@ -920,6 +955,101 @@ message ControllerServiceCapability {
 
 ### Node Service RPC
 
+#### `NodePublishDevice`
+
+A Node Plugin MUST implement this RPC call if it has `PUBLISH_UNPUBLISH_DEVICE` node capability.
+This RPC is called by the CO when a workload that wants to use the specified volume is placed (scheduled) on a node.
+The Plugin SHALL assume that this RPC will be executed on the node where the volume will be used.
+This RPC MUST be called by the CO once per node, per volume.
+If the corresponding Controller Plugin has `PUBLISH_UNPUBLISH_VOLUME` controller capability and the Node Plugin has `PUBLISH_UNPUBLISH_DEVICE`, then the CO MUST guarantee that this RPC is called after `ControllerPublishVolume` is called for the given volume on the given node and returns a success.
+
+This operation MUST be idempotent.
+If this RPC failed, or the CO does not know if it failed or not, it MAY choose to call `NodePublishDevice` again, or choose to call `NodeUnpublishDevice`.
+
+```protobuf
+message NodePublishDeviceRequest {
+  // The API version assumed by the CO. This is a REQUIRED field.
+  Version version = 1;
+
+  // The ID of the volume to publish. This field is REQUIRED.
+  string volume_id = 2;
+
+  // The CO SHALL set this field to the value returned by
+  // `ControllerPublishVolume` if the corresponding Controller Plugin
+  // has `PUBLISH_UNPUBLISH_VOLUME` controller capability, and SHALL be
+  // left unset if the corresponding Controller Plugin does not have
+  // this capability. This is an OPTIONAL field.
+  PublishVolumeInfo publish_volume_info = 3;
+
+  // The path to which the volume will be published. It MUST be an
+  // absolute path in the root filesystem of the process serving this
+  // request. The CO SHALL ensure uniqueness of mount_path per volume.
+  // This is a REQUIRED field.
+  string mount_path = 4;
```

Comment:
Maybe call this `global_mount_path` to mentally disambiguate it from the workload mount_path. And possibly renmae the `mount_path` in `NodePublishVolumeRequest` to `workload_mount_path`? Also `global_mount_path` should be passed in on `NodePublishVolumeRequest` as well.

Reply:
> Maybe call this `global_mount_path`

I find this problematic as it conveys more than a single volume. How about changing them to `mount_path` and `bind_mount_path`?

> Also `global_mount_path` should be passed in on `NodePublishVolumeRequest` as well.

Absolutely. That was a miss on my part. This is the largest change I've made to the spec up to this point and figured I'd miss something. Thank you for catching this!

Reply:
> I find this problematic as it conveys more than a single volume. How about changing them to `mount_path` and `bind_mount_path`?

Are you proposing
* `NodePublishDeviceRequest` `mount_path` -> `mount_path` (no change)
* `NodePublishVolumeRequest` `mount_path` -> `bind_mount_path`

That would make sense if all volume plugins implement volume `NodePublishDevice`, but for volumes that don't, `bind_mount_path` won't make sense.

I'd prefer either
* `NodePublishDeviceRequest(global_mount_path)`
* `NodePublishVolumeRequest(global_mount_path, mount_path)`

Or:
* `NodePublishDeviceRequest(device_mount_path)`
* `NodePublishVolumeRequest(device_mount_path, mount_path)`

Or:
* `NodePublishDeviceRequest(device_mount_path)`
* `NodePublishVolumeRequest(device_mount_path, workload_mount_path)`

Verbosity is ugly, I agree, but in this case I think it's worth 1) differentiating between the workload path and the device path, and 2) having an obvious link between the "global/device" mount_path passed to `NodePublishDeviceRequest` and `NodePublishVolumeRequest`.

Reply:
Hi @saad-ali,

FWIW, I already updated to `global_target_path` and left the existing `target_path` alone and was about to push the change when your comment notification arrived. Are you okay with this? I prefer minimal changes where possible, and this seems sufficient.

Reply:
Hi @saad-ali,

I just pushed `global_target_path`. Obviously it can be changed again as this discussion progresses.